### PR TITLE
Refactor(eos_designs): Better error messages for missing keys

### PR DESCRIPTION
**Pull request by gmuloc**
## Change Summary

Refactor: Better error messages for missing keys

## Component(s) name

`arista.avd.eos_designs`

### CODE REVIEW

File: ansible_collections/arista/avd/plugins/action/eos_designs_structured_config.py [hunk 1]
```diff
@@ -96,6 +97,8 @@ def run(self, tmp: Any = None, task_vars: dict | None = None) -> dict:
                 result=result,
                 templar=self.templar,
             )
+        except AristaAvdMissingVariableError as e:
+            raise AnsibleActionFail(message=str(e)) from e
         except Exception as error:
             raise AnsibleActionFail(message=str(error)) from error
```
Comment (ClausHolbechArista) [inline ansible_collections/arista/avd/plugins/action/eos_designs_structured_config.py]: Should we not add the host field here too?

File: python-avd/pyavd/_eos_designs/eos_designs_facts/vlans.py [hunk 2]
```diff
@@ -86,9 +86,9 @@ def _local_endpoint_vlans_and_trunk_groups(self: EosDesignsFacts) -> tuple[set,
         for connected_endpoints_key in self.shared_utils.connected_endpoints_keys:
             connected_endpoints = get(self._hostvars, connected_endpoints_key["key"], default=[])
             for connected_endpoint in connected_endpoints:
-                for adapter in connected_endpoint.get("adapters", []):
+                for index, adapter in enumerate(connected_endpoint.get("adapters", [])):
                     adapter_settings = self.shared_utils.get_merged_adapter_settings(
-                        adapter, context=f"{connected_endpoints_key['key']}[{connected_endpoint['name']}].adapters"
+                        adapter, context=f"{connected_endpoints_key['key']}[{connected_endpoint['name']}].adapters[{index}]"
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_eos_designs/eos_designs_facts/vlans.py]:
```suggestion
                        adapter, context=f"{connected_endpoints_key['key']}[name={connected_endpoint['name']}].adapters[{index}]"
```

File: python-avd/pyavd/_errors/__init__.py [hunk 3]
```diff
@@ -12,17 +12,44 @@ def _json_path_to_string(self, json_path: list[str | int]) -> str:
...
         for index, elem in enumerate(json_path):
             if isinstance(elem, int):
-                path += "[" + str(elem) + "]"
+                path += f"[{elem!s}]"
             else:
                 if index == 0:
                     path += elem
                     continue
-                path += "." + elem
+                path += f".{elem!s}"
```
Comment (ClausHolbechArista) [inline python-avd/pyavd/_errors/__init__.py]: !s is default for f-strings.

**Review by ClausHolbechArista — COMMENTED** [review]
This is a lot of changes. Seems like we most messages could be replaced by a "var_path" and a "context" or similar. So we could build the message with similar wording every time.

File: ansible_collections/arista/avd/plugins/action/eos_designs_structured_config.py (hunk 1 above)
Comment (gmuloc) [reply ansible_collections/arista/avd/plugins/action/eos_designs_structured_config.py]: it does not make sense for this part as we are already running it per host and so we are getting the proper error messages:


```
TASK [arista.avd.eos_designs : Generate device configuration in structured format] ***
ok: [legacy-autovpn-rr1 -> localhost]
ok: [legacy-autovpn-rr2 -> localhost]
ok: [legacy-autovpn-edge -> localhost]
ok: [legacy-autovpn-edge-no-default-policy -> localhost]
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: pyavd._errors.AristaAvdMissingVariableError: bgp_peer_groups.wan_overlay_peers.listen_range_prefixes
fatal: [cv-pathfinder-edge-no-default-policy -> localhost]: FAILED! => {"changed": false, "msg": "bgp_peer_groups.wan_overlay_peers.listen_range_prefixes"}
ok: [cv-pathfinder-edge-custom-default-policy -> localhost]
ok: [cv-pathfinder-edge2A -> localhost]
ok: [cv-pathfinder-edge2B -> localhost]
ok: [site-ha-enabled-leaf2A -> localhost]
ok: [site-ha-enabled-leaf2B -> localhost]
ok: [cv-pathfinder-transit1A -> localhost]
```

File: python-avd/pyavd/_eos_designs/eos_designs_facts/vlans.py (hunk 2 above)
Comment (gmuloc) [reply python-avd/pyavd/_eos_designs/eos_designs_facts/vlans.py]: thats not what we have in the other places where the message is so would need to change it there as well

File: python-avd/pyavd/_errors/__init__.py (hunk 3 above)
Comment (gmuloc) [reply python-avd/pyavd/_errors/__init__.py]: I blame ruff for suggesting it :)

File: ansible_collections/arista/avd/plugins/action/eos_designs_facts.py [hunk 4]
```diff
@@ -194,9 +194,12 @@ def render_avd_switch_facts(self, avd_switch_facts_instances: dict) -> dict:
...
             try:
                 rendered_facts[host] = {"switch": avd_switch_facts_instances[host]["switch"].render()}
-            except AristaAvdMissingVariableError as e:
-                msg = f"{e} is required but was not found for host '{host}'"
-                raise AnsibleActionFail(msg) from e
+            except AristaAvdError as e:
+                message = str(e)
+                if message.endswith("."):
+                    message = message[:-1]
+                msg = f"{message} for host '{host}'."
```
Comment (gmuloc) [inline ansible_collections/arista/avd/plugins/action/eos_designs_facts.py]: `remove_suffix(".")` (takes a tuple)

File: ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FABRIC_INVALID_UPLINK_PORT_CHANNEL_ID_2.yml [hunk 5]
```diff
@@ -42,5 +42,5 @@ l2leaf:
           id: 2
 
 expected_error_message: >-
-  Unexpected failure during module execution: 'uplink_switch_port_channel_id' must be between 1 and 2000 for MLAG switches.
+  'uplink_switch_port_channel_id' must be between 1 and 2000 for MLAG switches.
```
Comment (gmuloc) [inline ansible_collections/arista/avd/molecule/eos_designs_negative_unit_tests/inventory/group_vars/FABRIC_INVALID_UPLINK_PORT_CHANNEL_ID_2.yml]: why are we not printing an extra "for host" here?

**Review by ClausHolbechArista — APPROVED** [review]

**Review by carlbuchmann — APPROVED** [review]
LGTM

**Pull request merged**